(staging PR) server: add model management and proxy

ngxson opened the pull request:
The feature was implemented using subprocess, meaning there will be one main "router" server whose the job is to create other "child" processes that will actually run the inference.

This system was design and test against these unexpected cases:
- Child process suddenly exit due to error (for example, a `GGML_ASSERT`)
- Child process failed to load (for example, the system cannot launch the process)
- Router process suddenly exit due to error. In this case, child processes automatically stop themself.

These steps happen when user request the router to launch a model instance:
1. Check if the model already had a process; if yes, skip
2. Construct argv and envp to launch the child process; a random HTTP port is selected for each child process
3. Start the child process
4. Create a thread to read child's stdout/stderr and forward it to main process, with a prefix `[port_number]`
5. Inside child process, it notifies router server its "ready" status, then spawn a thread to monitor its stdin

If the child process exits, router server knows that as soon as stdout/stderr is closed; In reversed, if router process exits (or router request child process to stop), it simply close the stdin of the child. The child detects this and internally calls its SIGTERM/SIGINT handler.

---

Other changes included in the PR:
- Remove `DEFAULT_MODEL_PATH`
- If `-m, --model` is not specified, `common_params_parse_ex` will return an error (except for server)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Router mode: manage multiple models with load/unload/status endpoints and automatic routing
  * UI: model selector dropdown, detailed Model Information dialog, and improved audio recording controls (auto-mic)
  * Attachments: better handling/previews for audio, images, PDFs, and text; unified attachment UX

* **Documentation**
  * New guide for multi-model router setup and usage

* **Tests**
  * Added router streaming endpoint test suite

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ngxson committed:
server: add model management and proxy

ngxson committed:
fix compile error

ngxson committed:
does this fix windows?

ngxson committed:
fix windows build

ngxson committed:
use subprocess.h, better logging

ngxson committed:
add test

ngxson committed:
fix windows

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

allozaur committed:
feat: Model/Router server architecture WIP

ngxson committed:
more stable

ngxson committed:
fix unsafe pointer

ngxson committed:
also allow terminate loading model

ngxson committed:
add is_active()

allozaur committed:
refactor: Architecture improvements

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

ngxson committed:
tmp apply upstream fix

ngxson committed:
address most problems

ngxson committed:
address thread safety issue

ngxson committed:
address review comment

ngxson committed:
add docs (first version)

ngxson committed:
address review comment

allozaur committed:
feat: Improved UX for model information, modality interactions etc

allozaur committed:
chore: update webui build output

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

allozaur committed:
refactor: Use only the message data `model` property for displaying model used info

allozaur committed:
chore: update webui build output

ngxson committed:
add --models-dir param

allozaur committed:
feat: New Model Selection UX WIP

allozaur committed:
chore: update webui build output

allozaur committed:
feat: Add auto-mic setting

allozaur committed:
feat: Attachments UX improvements

ngxson committed:
implement LRU

ngxson committed:
remove default model path

ngxson committed:
better --models-dir

ngxson committed:
add env for args

ngxson committed:
address review comments

ngxson committed:
fix compile

allozaur committed:
refactor: Chat Form Submit component

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

ngxson committed:
ad endpoint docs

ngxson committed:
Merge remote-tracking branch 'webui/allozaur/server_model_management_v1_2' into xsn/server_model_maagement_v1_2

Co-authored-by: Aleksander <aleksander.grygier@gmail.com>

allozaur committed:
feat: Add copy to clipboard to model name in model info dialog

allozaur committed:
feat: Model unavailable UI state for model selector

allozaur committed:
feat: Chat Form Actions UI logic improvements

allozaur committed:
feat: Auto-select model from last assistant response

allozaur committed:
chore: update webui build output

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatAttachments/ChatAttachmentsList.svelte:
```diff
@@ -88,31 +91,30 @@
 					id: `attachment-${index}`,
 					name: attachment.name,
-					type: attachment.mimeType || 'audio',
+					type: 'application/pdf',
```
Use enum here

allozaur replied:
b7ba13b6a0418060cfbafd34a29ea2904feddc99

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatAttachments/ChatAttachmentsList.svelte:
```diff
@@ -2,10 +2,13 @@
 	import { getFileTypeCategory } from '$lib/utils/file-type';
+	import { FileTypeCategory } from '$lib/enums/files';
+	import { ModelModality } from '$lib/enums/model';
+	import { AttachmentType } from '$lib/enums/attachment';
```
Improve imports globally to use simply `$lib/enums`

allozaur replied:
ef5f9d07b0086726941a736b950b495dea994650

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatAttachments/ChatAttachmentsViewAll.svelte:
```diff
@@ -72,31 +75,30 @@
 					id: `attachment-${index}`,
 					name: attachment.name,
-					type: 'application/pdf',
+					type: 'text',
```
Enum

allozaur replied:
b7ba13b6a0418060cfbafd34a29ea2904feddc99

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormActions/ChatFormActions.svelte:
```diff
@@ -4,42 +4,51 @@
 	let {
 		canSend = false,
-		class: className = '',
+		className = '',
```
```suggestion
		class?: string;
		disabled?: boolean;
		isLoading?: boolean;
		isRecording?: boolean;
		hasText?: boolean;
		uploadedFiles?: ChatUploadedFile[];
		onFileUpload?: (fileType?: FileTypeCategory) => void;
		onMicClick?: () => void;
		onStop?: () => void;
	}

	let {
		canSend = false,
		class: className = '',
```

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormActions/ChatFormActions.svelte:
```diff
@@ -51,15 +60,19 @@
-			<span class="sr-only">Send</span>
-			<ArrowUp class="h-12 w-12" />
-		</Button>
+		{#if shouldShowSubmitButton}
```
```suggestion
		{:else}
```

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:
```diff
@@ -244,21 +246,52 @@
-				</button>
+						{tokensPerSecond.toFixed(2)} tokens/s
+					</span>
+				{/if}
```
Componentize

allozaur replied:
f8ff39c64ed39eda18ecaa51f9831dd81e6c8ff8

allozaur commented on tools/server/webui/src/lib/components/app/dialogs/DialogChatAttachmentPreview.svelte:
```diff
@@ -32,16 +37,17 @@
 				? 'image'
-				: attachment?.type === 'textFile'
+				: attachment?.type === AttachmentType.TEXT
 					? 'text'
```
Enum

ngxson committed:
expose args and exit_code in API

allozaur commented on tools/server/webui/src/lib/components/app/dialogs/DialogModelInformation.svelte:
```diff
@@ -0,0 +1,245 @@
+	let { open = $bindable(), onOpenChange }: Props = $props();
+
+	let serverProps = $derived(serverStore.serverProps);
+	let modalities = $derived(serverStore.supportedModalities);
```
NEeds to be updated accordingly with the router server updates

allozaur commented on tools/server/webui/src/lib/components/app/dialogs/DialogModelInformation.svelte:
```diff
@@ -0,0 +1,245 @@
+	function formatNumber(num: number | unknown): string {
+		if (typeof num !== 'number') return 'Unknown';
+		return num.toLocaleString();
+	}
```
Move these functions to `$lib/utils`

allozaur replied:
41764b8fa050e80fb247d9b7f7b9625bc5830017

ngxson committed:
add note

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:
```diff
@@ -244,21 +246,52 @@
-					<span>Model used:</span>
-				</span>
+						<Copy class="ml-1 h-3 w-3 " />
+					</button>
```
Use `BadgeModelName.svelte` component

allozaur replied:
16747dee5b0fe0a652e5250e1a632287fae4b15b

allozaur commented on tools/server/webui/src/lib/components/app/misc/BadgeModelName.svelte:
```diff
@@ -0,0 +1,53 @@
+			</Tooltip.Trigger>
+
+			<Tooltip.Content>
+				{onclick ? 'Click for model details' : 'Model name'}
```
```suggestion
				{onclick ? 'Click for model details' : model}
```

allozaur replied:
39fb1c2b176a60bc679a7538ee6ded3b134f6a6f

ngxson committed:
support extra_args on loading model

ngxson committed:
allow reusing args if auto_load

ngxson committed:
typo docs

allozaur commented on tools/server/webui/src/lib/services/chat.ts:
```diff
@@ -722,6 +711,33 @@ export class ChatService {
+			throw error;
+		}
+	}
+
```
This should not be in the Chat service, but a new one — `router.ts` / `models.ts` / `server.ts` (not sure which path i will go yet)

allozaur commented on tools/server/webui/src/lib/stores/server.svelte.ts:
```diff
@@ -52,6 +55,10 @@ class ServerStore {
+	private _serverMode = $state<ServerMode | null>(null);
+	private _selectedModel = $state<string | null>(null);
+	private _availableModels = $state<ApiRouterModelMeta[]>([]);
+	private _modelLoadingStates = $state<Map<string, boolean>>(new Map());
```
Need to rethink this if it should go to a separate service + store (`router`/`models`)

allozaur commented on tools/server/README.md:
@ngxson maybe it's worth adding info about auto-loading that happens when we simply do a request to `/chat/completions` with `model` param specified in the payload and how it's behaving if not handled manually with `/load` / `/unload` endpoints?

allozaur reviewed (commented):
Mostly notes/remarks to myself + just one question regarding the README

ngxson committed:
oai-compat /models endpoint

ngxson committed:
cleaner

ngxson commented:
@coderabbitai review only C++ code

ngxson committed:
address review comments

allozaur committed:
feat: Use `model` property for displaying the `repo/model-name` naming format

allozaur committed:
refactor: Attachments data

allozaur committed:
chore: update webui build output

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatAttachments/ChatAttachmentsList.svelte:
```diff
@@ -88,31 +91,30 @@
 					id: `attachment-${index}`,
 					name: attachment.name,
-					type: 'application/pdf',
+					type: 'text',
```
b7ba13b6a0418060cfbafd34a29ea2904feddc99

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatAttachments/ChatAttachmentsViewAll.svelte:
```diff
@@ -72,31 +75,30 @@
 					id: `attachment-${index}`,
 					name: attachment.name,
-					type: attachment.mimeType || 'audio',
+					type: 'application/pdf',
```
b7ba13b6a0418060cfbafd34a29ea2904feddc99

allozaur committed:
refactor: Enum imports

allozaur committed:
feat: Improve Model Selector responsiveness

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Cleanup

allozaur commented on tools/server/webui/src/lib/components/app/dialogs/DialogChatAttachmentPreview.svelte:
```diff
@@ -32,16 +37,17 @@
+		uploadedFile
+			? uploadedFile.type
+			: attachment?.type === AttachmentType.IMAGE
 				? 'image'
```
b7ba13b6a0418060cfbafd34a29ea2904feddc99

allozaur committed:
refactor: Cleanup

allozaur committed:
refactor: Formatters

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Copy To Clipboard Icon component

allozaur committed:
chore: update webui build output

allozaur commented on tools/server/webui/src/lib/components/app/dialogs/DialogModelInformation.svelte:
```diff
@@ -0,0 +1,245 @@
+	import { ChatService } from '$lib/services/chat';
+	import type { ApiModelListResponse } from '$lib/types/api';
+	import { Copy } from '@lucide/svelte';
+	import { copyToClipboard } from '$lib/utils/copy';
```
e92ce0791667b452980b6fb1b5198046a3e83674

allozaur committed:
refactor: Cleanup

allozaur committed:
chore: update webui build output

allozaur commented on tools/server/webui/src/lib/components/app/misc/SelectorModel.svelte:
```diff
@@ -244,63 +207,85 @@
+		if (currentModel) {
+			return options.find((option) => option.model === currentModel);
+		}
 		if (activeId) {
```
39fb1c2b176a60bc679a7538ee6ded3b134f6a6f

allozaur commented on tools/server/webui/src/lib/components/app/misc/SelectorModel.svelte:
```diff
@@ -244,63 +207,85 @@
+					capabilities: [] // Empty array for single model mode
+				};
+			}
+			return undefined;
```
39fb1c2b176a60bc679a7538ee6ded3b134f6a6f

allozaur committed:
refactor: UI badges

allozaur committed:
chore: update webui build output

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

allozaur committed:
refactor: Cleanup

allozaur committed:
refactor: Cleanup

allozaur committed:
chore: update webui build output

ngxson committed:
add --models-allow-extra-args for security

ngxson committed:
nits

ngxson committed:
add stdin_file

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

ngxson committed:
fix merge

allozaur committed:
fix: Retrieve lost setting after resolving merge conflict

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

allozaur committed:
refactor: DatabaseStore -> DatabaseService

allozaur committed:
refactor: Database, Conversations & Chat services + stores architecture improvements (WIP)

allozaur committed:
refactor: Remove redundant settings

allozaur committed:
refactor: Multi-model business logic WIP

allozaur committed:
chore: update webui build output

allozaur committed:
feat: Switching models logic for ChatForm or when regenerating messges + modality detection logic

allozaur committed:
chore: update webui build output

allozaur committed:
fix: Add `untrack` inside chat processing info data logic to prevent infinite effect

allozaur committed:
fix: Regenerate

allozaur committed:
feat: Remove redundant settigns + rearrange

allozaur committed:
fix: Audio attachments

allozaur committed:
refactor: Icons

allozaur committed:
chore: update webui build output

allozaur committed:
feat: Model management and selection features WIP

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Improve server properties management

allozaur committed:
refactor: Icons

allozaur committed:
chore: update webui build output

allozaur committed:
feat: Improve model loading/unloading status updates

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Improve API header management via utility functions

ngxson committed:
remove support for extra args

ngxson committed:
set hf_repo/docker_repo as model alias when posible

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

allozaur committed:
refactor: Remove ConversationsService

allozaur committed:
refactor: Chat requests abort handling

allozaur committed:
refactor: Server store

ngxson committed:
tmp webui build

allozaur committed:
refactor: Model modality handling

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Processing state reactivity

allozaur committed:
fix: UI

allozaur committed:
refactor: Services/Stores syntax + logic improvements

Refactors components to access stores directly instead of using exported getter functions.

This change centralizes store access and logic, simplifying component code and improving maintainability by reducing the number of exported functions and promoting direct store interaction.

Removes exported getter functions from `chat.svelte.ts`, `conversations.svelte.ts`, `models.svelte.ts` and `settings.svelte.ts`.

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into allozaur/server_model_management_v1_2

allozaur committed:
refactor: Architecture cleanup

allozaur committed:
feat: Improve statistic badges

allozaur committed:
feat: Condition available models based on modality + better model loading strategy & UX

allozaur committed:
docs: Architecture documentation

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

allozaur committed:
feat: Update logic for PDF as Image

ngxson committed:
add TODO for http client

allozaur committed:
refactor: Enhance model info and attachment handling

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Components naming

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Cleanup

allozaur committed:
refactor: DRY `getAttachmentDisplayItems` function + fix UI

allozaur committed:
chore: update webui build output

allozaur committed:
fix: Modality detection improvement for text-based PDF attachments

allozaur committed:
refactor: Cleanup

allozaur committed:
docs: Add info comment

allozaur committed:
refactor: Cleanup

allozaur committed:
re

allozaur committed:
refactor: Cleanup

allozaur committed:
refactor: Cleanup

allozaur committed:
feat: Attachment logic & UI improvements

allozaur committed:
refactor: Constants

allozaur committed:
feat: Improve UI sidebar background color

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Utils imports + move types to `app.d.ts`

allozaur committed:
test: Fix Storybook mocks

allozaur committed:
chore: update webui build output

allozaur committed:
Merge branch 'master' into allozaur/server_model_management_v1_2

allozaur committed:
test: Update Chat Form UI tests

allozaur committed:
refactor: Tooltip Provider from core layout

allozaur committed:
refactor: Tests to separate location

allozaur committed:
Merge remote-tracking branch 'origin/allozaur/server_model_management_v1_2' into allozaur/server_model_management_v1_2

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

ngxson committed:
decouple server_models from server_routes

allozaur committed:
test: Move demo test  to tests/server

allozaur committed:
refactor: Remove redundant method

allozaur committed:
chore: update webui build output

ngxson committed:
also route anthropic endpoints

ngxson committed:
Merge remote-tracking branch 'webui/allozaur/server_model_management_v1_2' into xsn/server_model_management_v1_2

ngxson committed:
fix duplicated arg

ngxson committed:
fix invalid ptr to shutdown_handler

ggerganov committed:
server : minor

ngxson committed:
rm unused fn

ngxson committed:
add ?autoload=true|false query param

ngxson committed:
Merge branch 'master' into xsn/server_model_management_v1_2

allozaur committed:
refactor: Remove redundant code

allozaur committed:
Merge remote-tracking branch 'ngxson/xsn/server_model_management_v1_2' into xsn/server_model_management_v1_2

allozaur committed:
docs: Update README documentations + architecture & data flow diagrams

allozaur committed:
fix: Disable autoload on calling server props for the model

allozaur committed:
chore: update webui build output

ngxson committed:
fix ubuntu build

allozaur committed:
fix: Model status reactivity

allozaur committed:
fix: Modality detection for MODEL mode

allozaur committed:
chore: update webui build output
